src/libasr/codegen/llvm_utils.cpp

```diff
@@ -1245,11 +1245,26 @@ namespace LFortran {
         return LLVM::CreateLoad(*builder, value_ptr);
     }
 
-    llvm::Value* LLVMList::read_item(llvm::Value* list, llvm::Value* pos, bool get_pointer,
+    llvm::Value* LLVMList::get_list_pos(llvm::Value* list, llvm::Value* pos) {
+        llvm::Value* list_len = LLVMList::len(list);
+
+        llvm::Type* type = llvm::Type::getInt32Ty(context);
+        llvm::AllocaInst *tmp_pos = builder->CreateAlloca(type, nullptr);
+        llvm::Value* zero = llvm::ConstantInt::get(llvm::Type::getInt32Ty(context),
+                                            llvm::APInt(32, 0));
+        LLVM::CreateStore(*builder,
+            builder->CreateSRem(builder->CreateAdd(pos, list_len), list_len),
+            tmp_pos);
```

Comment:
In here we do `(pos+list_len) % list_len`. That is much slower than just `pos`. So I think we should not allow this in ASR, I think we should always assume in `ListItem` has index between `0` and `size-1`.

In Debug mode we'll add a runtime check for this.

For things like `x[-1]` or `x[-5]` the frontend will create ASR of the form `x[len(x)-1]` and `x[len(x)-5]`. So that would still work.

Finally, for `x[i]` where `i=-5` I think we should simply not allow such usage. I think I might have never used this in practice anyway, so our (future) runtime Debug check would simply give an error message, and the user would have to fix the code, in this case `x[i]` -> `x[len(x)+i]` if `i < 0`.

If the user absolutely requires this usage, he can do something like `x[(i+len(x)) % len(x)]`, that way the behavior is as intended and it is explicit what is going on, and (most importantly!) we are not losing any performance.

Reply:
Correct. Note that figuring out whether a variable is negative or not is a runtime operation. For some cases you can figure out in compile time (like for constant negative integers as cited above). Now the problem is allowing negative indexing as `x[-5]` but not via `x[i]` where `i = -5` is simply a source of confusion. For example I will be able to do the following,

```python
if flag:
    x[-5] = 0
else:
    x[-20] = 0
```

but not the following,

```python
if flag:
    i = -5
else:
    i = -20
x[i] = 0
```

Even though both codes are semantically mean the same. I would say add an option to allow negative indexing and generate code in backends accordingly. Explain to the users in the documentation that how negative indexing degrades performance, hence we don't support it by default.

Either allow negative indexing completely or don't allow it.

Reply:
> Either allow negative indexing completely or don't allow it.

That would be ideal, but I don't know how to enable it for general expressions like "i", since we only know at runtime if it is positive or negative. So for performance reasons I would only allow it for compile time constant integer expressions.